src/wallet/wallet.cpp

```diff
@@ -3346,20 +2914,17 @@ DBErrors CWallet::LoadWallet(bool& fFirstRunRet)
     {
         if (database->Rewrite("\x04pool"))
         {
-            setInternalKeyPool.clear();
-            setExternalKeyPool.clear();
-            m_pool_key_to_index.clear();
-            // Note: can't top-up keypool here, because wallet is locked.
-            // User will be prompted to unlock wallet the next operation
-            // that requires a new key.
+            for (const auto& spk_man_pair : m_spk_managers) {
+                spk_man_pair.second->RewriteDB();
+            }
         }
     }
 
-    {
-        LOCK(cs_KeyStore);
-        // This wallet is in its first run if all of these are empty
-        fFirstRunRet = mapKeys.empty() && mapCryptedKeys.empty() && mapWatchKeys.empty() && setWatchOnly.empty() && mapScripts.empty()
-            && !IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && !IsWalletFlagSet(WALLET_FLAG_BLANK_WALLET);
+    // This wallet is in its first run if all of these are empty and this isn't blank or no privkeys
+    fFirstRunRet = m_spk_managers.empty() && !IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && !IsWalletFlagSet(WALLET_FLAG_BLANK_WALLET);
```

Comment:
I wasn't sure if the new `m_spk_managers.empty()` condition here equivalent to the previous `mapKeys.empty() && mapCryptedKeys.empty() && mapWatchKeys.empty() && setWatchOnly.empty() && mapScripts.empty()` condition. I think it is equivalent because the `LoadWallet(*this)` call above will only call `GetLegacyScriptPubKeyMan()` internally if keys are found, and if it does not, `SetupLegacyScriptPubKeyMan()` will not have been called yet and  `m_spk_managers` will be empty. But correct me if I'm wrong.

I guess it's a little confusing that `GetLegacyScriptPubKeyMan` right will actually create a legacy keyman if one doesn't exist, and never return null, even though there are checks for null throughout the code. Maybe there should be two different methods for clarity:

```c++
LegacyScriptPubKeyMan* GetLegacyScriptPubKeyMan();
LegacyScriptPubKeyMan& GetOrCreateLegacyScriptPubKeyMan();
```

Just a suggestion, though, and correct me if I missed anything.

Reply:
Yes, that is correct. That's pretty much the only reason `GetLegacyScriptPubKeyMan` also creates a `LegacyScriptPubKeyMan`.

The checks for null are for the future when we have more `ScriptPubKeyMan` classes.

Reply:
I've added `GetOrCreateLegacyScriptPubKeyMan`